core/ir/synthetic/decode.c

```diff
@@ -0,0 +1,139 @@
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of VMware, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* decode.c -- a decoder for DR synthetic IR */
+
+#include "decode.h"
+
+#include "../globals.h"
+#include "encoding_common.h"
+#include "instr_api.h"
+#include "opnd_api.h"
+
+/* Decodes the raw bytes of an encoded instruction \p encoded_instr into DR instruction
+ * representation \p instr.
+ * Returns next instruction's pc.
+ *
+ * The smallest encoded instruction (i.e., with no operands) has 4 bytes and follows
+ * this scheme:
+ * |----------------------| |--| |----| |----|
+ * 31..               ..10  9,8   7..4   3..0
+ *       category          eflags #src  #dst
+ *
+ * 22 bits, category: holds values of type #dr_instr_category_t. Note that an instruction
+ * can belong to more than one category.
+ * 2 bits, eflags: most significant bit set to 1 indicates the instruction reads at least
+ * one arithmetic flag, least significant bit set to 1 indicates the instruction writes
+ * at least one arithmetic flag.
+ * 4 bits, #src: number of source operands (read) that are registers.
+ * 4 bits, #dst: number of destination operands (written) that are registers.
+ * (Note that we are only interested in register dependencies, hence operands that are
+ * not register, such as immediates or memory references, are not present)
+ *
+ * Instructions with operands can reach up to 14 bytes: the 4 bytes described above, and
+ * up to 10 register operands of 1 byte each.
+ * For example, an instruction with 4 operands (2 src, 2 dst) has 4 additional bytes that
+ * are encoded following this scheme:
+ * |--------| |--------| |--------| |--------|
+ * 31.. ..24  23.. ..16  15..  ..8  7..   ..0
+ * src_opnd0  src_opnd1  dst_opnd0  dst_opnd1
+ *
+ * We assume all encoded values to be little-endian.
```

Comment:
It is not clear enough how this works: the examples imply things are broken into 4-byte little-endian words.  But what about operand counts that are not a multiple of 4?  What happens then?  Are the rest of the bits in the word 0, or undefined, or is it no longer a word at that point?

Should it be broken into words, or just a stream of bytes for the operands? 

If the max operand count is indeed 8, and we split into 4-byte words, then we always have the initial word, followed by either 1 or 2 words with the operands?

If it is broken in 4-byte words: are instruction PC values always aligned to 4?  Should our decode and encode routines enforce that?

Reply:
This is documented in core/ir/synthetic/encoding_common.h.